docs/proposals/20220221-runtime-SDK.md

```diff
@@ -0,0 +1,802 @@
+    // output types to a catalog.
+    AddToCatalog = catalogBuilder.AddToCatalog
+
+    // localSchemeBuilder provides access to the SchemeBuilder used for managing Runtime Hooks
+    // input and output types defined in this package.
+    // NOTE: this object is required to allow registration of automatically generated
+    // conversions func.
+    localSchemeBuilder = catalogBuilder.SchemeBuilder
+)
+
+func init() {
+    // Add Open API definitions for Runtime Hooks input and output types in this package
+    // NOTE: the GetOpenAPIDefinitions func is automatically generated by openapi-gen.
+    catalogBuilder.OpenAPIDefinitions(GetOpenAPIDefinitions)
+
+    // Register Runtime Hooks defined in this package and their input and output types.
+    catalogBuilder.RegisterHook(&BeforeUgradeHook{}, &BeforeUgradeInput{}, &BeforeUgradeOutput{})
+}
+```
+
+Given the above definitions, a catalog can finally be created as follow:
+
+```go
+var c = catalog.NewCatalog()
+
+func init() {
+    v1alpha1.AddToCatalog(c)
+    v1alpha2.AddToCatalog(c)
+    v1alpha3.AddToCatalog(c)
+}
+```
+
+The catalog provides the core knowledge required to manage all the Runtime Hooks supported by Cluster API;
+the first application of such knowledge will be to retrieve all the info required to generate the OpenAPI specification
+for Runtime Hooks with a dedicated tool under `hack/tools`.
+
+### Discovering Runtime Extensions
+
+_Note: the controller described in this paragraph will be executed only if the Runtime Extension feature flag is set to true._
+
+Cluster API is going to implement a new controller that looks at Runtime Extension Configurations; the mai
+responsibility of this controller should be to maintain an internal, shared **registry** of available extensions
+at a given time.
+
+Please note that the Runtime Extensions registry also provides a single point to centralize a set of common behaviors
+supporting interaction with those external components, thus making the adoption of this feature scalable - 
+in the sense of being used for an increasing numbers of use cases in Cluster API - while operating consistently
+across the board.
+
+A first behavior that falls into this category is the implementation of exponential backoff mechanisms
+in case of errors, thus preventing Cluster API from creating pressure on HTTP Servers recovering from or with
+ongoing operational issues.
+
+Another cross-cutting concern is about ensuring that Runtime Extensions, which are external components triggered
+in the middle of Cluster API controllers logic, do not block the reconciliation process indefinitely 
+(e.g by enforcing a maximum timeout for all the Runtime Extensions calls).
+
+### Calling Runtime Extensions
+
+_Note: the code described in this paragraph will be executed only if the Runtime Extension feature flag is set to true._
+
+Cluster API is going to implement calls to registered Runtime Extensions at well-known moments of the Cluster’s lifecycle.
```

Comment:
```suggestion
Cluster API is going to implement calls to registered Runtime Extensions at well-known moments.
```
Maybe we can just simplify a bit here, and by that keep the door a bit wider open to non-lifecycle-related hooks

Reply:
Same consideration as above, let's focus on the main use case for the sake of simplicity